operator/pkg/upgrade/upgrade.go

```diff
@@ -2,411 +2,157 @@ package upgrade
-			desired.Spec.Extensions.Helm.Charts[i] = operatorChart
-			return desired
+	// create an autopilot upgrade plan if one does not yet exist
+	var plan apv1b2.Plan
+	okey := client.ObjectKey{Name: "autopilot"}
+	if err := cli.Get(ctx, okey, &plan); err != nil && !errors.IsNotFound(err) {
+		return fmt.Errorf("failed to get upgrade plan: %w", err)
+	} else if errors.IsNotFound(err) {
+		// if the kubernetes version has changed we create an upgrade command
+		fmt.Printf("Starting k0s autopilot upgrade plan to version %s\n", desiredVersion)
+
+		// there is no autopilot plan in the cluster so we are free to
+		// start our own plan. here we link the plan to the installation
+		// by its name.
+		if err := StartAutopilotUpgrade(ctx, cli, in, meta); err != nil {
+			return fmt.Errorf("failed to start upgrade: %w", err)
 		}
 	}
-	desired.Spec.Extensions.Helm.Charts = append(desired.Spec.Extensions.Helm.Charts, operatorChart)
-	return desired
-}
-
-func getExistingClusterConfig(ctx context.Context, cli client.Client) (*k0sv1beta1.ClusterConfig, error) {
-	clusterConfig := &k0sv1beta1.ClusterConfig{}
-	err := cli.Get(ctx, client.ObjectKey{Name: clusterConfigName, Namespace: clusterConfigNamespace}, clusterConfig)
-	if err != nil {
-		return nil, fmt.Errorf("get chart: %w", err)
-	}
-	return clusterConfig, nil
-}
-
-func getOperatorChart(ctx context.Context, cli client.Client, in *clusterv1beta1.Installation) (*k0sv1beta1.Chart, error) {
-	metadata, err := release.MetadataFor(ctx, in, cli)
-	if err != nil {
-		return nil, fmt.Errorf("get release metadata: %w", err)
-	}
 
-	// fetch the current clusterConfig
-	var clusterConfig k0sv1beta1.ClusterConfig
-	err = cli.Get(ctx, client.ObjectKey{Name: "k0s", Namespace: "kube-system"}, &clusterConfig)
-	if err != nil {
-		return nil, fmt.Errorf("get cluster config: %w", err)
+	// restart this function/pod until the plan is complete
+	if !autopilot.HasThePlanEnded(plan) {
+		return fmt.Errorf("an autopilot upgrade is in progress (%s)", plan.Spec.ID)
 	}
 
-	combinedConfigs, err := charts.K0sHelmExtensionsFromInstallation(ctx, in, metadata, &clusterConfig)
-	if err != nil {
-		return nil, fmt.Errorf("failed to get helm charts from installation: %w", err)
+	if autopilot.HasPlanFailed(plan) {
+		reason := autopilot.ReasonForState(plan)
+		return fmt.Errorf("autopilot plan failed: %s", reason)
 	}
 
-	cfgs := &k0sv1beta1.HelmExtensions{}
-	cfgs, err = v1beta1.ConvertTo(*combinedConfigs, cfgs)
-	if err != nil {
-		return nil, fmt.Errorf("convert to k0s helm type: %w", err)
+	// check if this was actually a k0s upgrade plan, or just an image download plan
+	isK0sUpgrade := false
+	for _, command := range plan.Spec.Commands {
+		if command.K0sUpdate != nil {
+			isK0sUpgrade = true
+			break
+		}
 	}
-
-	for _, chart := range cfgs.Charts {
-		if chart.Name == operatorChartName {
-			return &chart, nil
+	// if this was not a k0s upgrade plan, we can just delete the plan and restart the function to get a k0s upgrade
```

Comment:
should this be done before checking if the plan has ended because it might've been a non-k0s upgrade plan?

Reply:
I think that even if the plan that's in progress isn't a k0s upgrade plan we should still wait for it to complete
similarly if it failed we should probably avoid messing things up further